# DP: diff between the standalone branch and the forge branch
# DP: (note that master has no delta against standalone)

diff --git a/post-receive-email b/post-receive-email
index abde17e..0278576 100755
--- a/post-receive-email
+++ b/post-receive-email
@@ -8,6 +8,11 @@
 #	Thorsten “mirabilos” Glaser <t.glaser@tarent.de>
 # Copyright © 2016
 #	martin f krafft (madduck)
+# Part of FusionForge (in particular Evolvis).
+#
+# This script can only be used in EvolvisForge configured git
+# repositories for a (forge) project. It accesses the forge
+# configuration and uses Evolvis-specific data files.
 #
 # Copyright (c) 2007 Andy Parkins
 #
@@ -64,23 +69,26 @@
 # Config
 # ------
 # hooks.mailinglist
-#   This is the list that all pushes will go to; leave it blank to not send
-#   emails for every ref update. This fills the To header, so use comma for
-#   multiple recipients (but please only use addresses, not full names).
+#   This is the list that all pushes will go to; leave it blank to send all
+#   eMails to the forge commits list. This fills the To header, so use comma
+#   for multiple recipients (but please only use addresses, not full names).
 # hooks.announcelist
 #   This is the list(s) that all pushes of annotated tags will go to. Leave
 #   it blank to default to the mailinglist field. The announce emails list
 #   the short log summary of the changes since the last annotated tag.
 # hooks.replyto
 #   If set, replies are requested to be directed to this eMail address(es).
+#   This defaults to the forge discuss list.
 # hooks.envelopesender
 #   If set then the -f option is passed to sendmail to allow the envelope
 #   sender to be set (one address only). This is also used as From header.
+#   This defaults to your user eMail address at the forge.
 # hooks.emailprefix
 #   If set, e.g. to '[SCM] ', all eMails have their subjects prefixed
 #   with this string to aid filtering.
 # hooks.shortdesc
-#   If set, this is used in the Subject instead of the description file.
+#   If set, this is used in the Subject to prefix messages instead of
+#   the forge subrepository name in curly braces.
 # hooks.showrev
 #   The shell command used to format each revision in the email, with
 #   "%s" replaced with the commit id.  Defaults to "git rev-list -1
@@ -129,6 +137,44 @@ export LANG=C LC_ADDRESS=C LC_COLLATE=C LC_CTYPE=C.UTF-8 \
     LC_MONETARY=C LC_NAME=C LC_NUMERIC=C LC_PAPER=C \
     LC_TELEPHONE=C LC_TIME=C
 
+# ---------------------------- Forge Integration
+
+# get configuration values
+keys=domain_name:lists_host:sys_name
+while IFS= read -r line; do
+	[[ $line != *' '* || $line = '#'* ]] && continue
+	key=${line%% *}
+	value=${line#* }
+	[[ :$keys: = *:$key:* ]] || continue
+	eval $key=\$value
+done </etc/gforge/shell.inc
+
+# find out where we are
+whereami() {
+	echo >&2 "Where am I? GIT_DIR is not set, or this is not"
+	echo >&2 "an Evolvis gforge-plugin-scmgit repository!"
+	exit 1
+}
+[[ -n $GIT_DIR ]] || whereami
+GIT_DIR=$(realpath "$GIT_DIR")
+[[ -n $GIT_DIR ]] || whereami
+parentdir=$GIT_DIR
+max_recursion=64
+while [[ $parentdir != / ]]; do
+	(( --max_recursion )) || whereami
+	parentdir=$(realpath "$parentdir/..")
+	[[ -s $parentdir/ffgroup.nfo ]] && break
+done
+[[ -s $parentdir/ffgroup.nfo ]] || whereami
+[[ ${parentdir##*/} = "$(<$parentdir/ffgroup.nfo)" ]] || whereami
+reponame=${GIT_DIR#"$parentdir/"}
+
+ff_group=${parentdir##*/}
+ff_subrepo=${reponame%.git}
+ff_sender=$(id -un)@${domain_name}
+ff_target=${ff_group}-commits@$lists_host
+ff_replyto=${ff_group}-discuss@$lists_host
+
 # ---------------------------- Functions
 
 #
@@ -369,6 +415,7 @@ generate_email_footer()
 	hooks/post-receive
 	--${SPACE}
 	$projectdesc
+	($sys_name project $ff_group repository $ff_subrepo)
 	EOF
 }
 
@@ -861,8 +908,10 @@ then
 fi
 
 recipients=$(git config hooks.mailinglist)
+: "${recipients:=$ff_target}"
 announcerecipients=$(git config hooks.announcelist)
 envelopesender=$(git config hooks.envelopesender)
+: "${envelopesender:=$ff_sender}"
 emailprefix=$(git config hooks.emailprefix)
 custom_showrev=$(git config hooks.showrev)
 maxlines=$(git config hooks.emailmaxlines)
@@ -870,8 +919,9 @@ diffopts=$(git config hooks.diffopts)
 : "${diffopts:="-p --stat --summary --find-copies-harder"}"
 kgbconffile=$(git config hooks.kgbconffile)
 replyto=$(git config hooks.replyto)
+: "${replyto:=$ff_replyto}"
 shortdesc=$(git config hooks.shortdesc)
-: "${shortdesc:=$projectdesc}"
+[[ -n $shortdesc ]] || shortdesc='{'$ff_subrepo'}'
 
 # --- Main loop
 # Allow dual mode: run from the command line just like the update hook, or
